### PR TITLE
Don't copy forwarded items

### PR DESCRIPTION
**Pull request by Tyilo**
Instead we store the items forwarded from nodes in that node and forward the node token of that node, which we can use to fetch the item with.
When forwarding from a pipeline, we now store that in the nodemap and fetch on nodes has been modified to also look for a key there.

As a result of this, you can now forward and fetch `unique_ptr`'s.

### CODE REVIEW

File: tpie/pipelining/node.cpp [hunk 1]
```diff
@@ -210,33 +210,56 @@ void node::forward_any(std::string key, boost::any value, memory_size_type k) {
...
-		ss << "Tried to fetch nonexistent key '" << key << '\'';
+		ss << "Tried to fetch nonexistent key '" << key
+		   << "' of type " << type_name
+		   << " in " << get_name() << " of type " << typeid(*this).name();
 		throw invalid_argument_exception(ss.str());
 	}
+
+	node_token::id_t fetch_from_id = it->second;
+	bits::node_map::ptr nodeMap = get_node_map()->find_authority();
+	node *fetch_from = nodeMap->get(fetch_from_id);
```
Comment (antialize) [inline tpie/pipelining/node.cpp]: We could, and prob should use auto for these three types.

File: tpie/pipelining/node.h [hunk 2]
```diff
@@ -573,46 +573,42 @@ class node {
...
-		return m_values.count(key) != 0;
+		return m_forwardedToHere.count(key) != 0;
 	}
 
 	///////////////////////////////////////////////////////////////////////////
 	/// \brief Fetch piece of auxiliary data as boost::any (the internal
 	/// representation).
 	///////////////////////////////////////////////////////////////////////////
-	boost::any fetch_any(std::string key);
+	boost::any fetch_any(std::string key, const std::string & type_name = "boost::any");
```
Comment (antialize) [inline tpie/pipelining/node.h]: The typename here is only used for error messages. I do not think it is worth the complications.

File: tpie/pipelining/node.cpp [hunk 3]
```diff
@@ -226,40 +226,42 @@ void node::add_forwarded_data(std::string key, node_token::id_t from_node) {
...
 
 	node_token::id_t fetch_from_id = it->second;
-	bits::node_map::ptr nodeMap = get_node_map()->find_authority();
 	node *fetch_from = nodeMap->get(fetch_from_id);
 	if (!fetch_from) {
+		return maybeany_t();
+	}
+
+	return fetch_from->get_forwarded_data_maybe(key);
+}
```
Comment (antialize) [inline tpie/pipelining/node.cpp]: A any type allready has a buildin unset value, and an operator bool() as well, so we do not need to wrap this in an optional i think. We can just return a any_noncopyable, howerever we would need a global empty any_noncopyable somewhere to return a reference to, so perhaps that would be just as bad..

File: tpie/pipelining/tokens.h [hunk 4]
```diff
@@ -193,6 +197,17 @@ class node_map {
...
+		m_pipelineForwards[key] = std::move(value);
+	}
+
+	maybeany_t fetch_maybe(std::string key) {
+		auto it = m_pipelineForwards.find(key);
+		if (it == m_pipelineForwards.end()) {
+			return maybeany_t();
+		}
+		return maybeany_t(it->second);
+	}
```
Comment (antialize) [inline tpie/pipelining/tokens.h]: who inserts into  m_pipelineForwards except ppl who call forward, what is the purpose of this method?

File: tpie/pipelining/tokens.cpp [hunk 5]
```diff
@@ -50,6 +50,9 @@ void node_map::link(node_map::ptr target) {
 	for (relmapit i = target->m_relationsInv.begin(); i != target->m_relationsInv.end(); ++i) {
 		m_relationsInv.insert(*i);
 	}
+	for (auto i = target->m_pipelineForwards.begin(); i != m_pipelineForwards.end(); ++i) {
+		m_pipelineForwards[i->first] = std::move(i->second);
+	}
```
Comment (antialize) [inline tpie/pipelining/tokens.cpp]: I see now why the hasmap is needed.

File: tpie/pipelining/container.h [hunk 6]
```diff
@@ -193,6 +193,12 @@ class any_noncopyable {
 	friend T & any_cast(any_noncopyable & a);
 	
 	friend void swap(any_noncopyable & l, any_noncopyable & r);
+
+	const std::type_info & type() const {
+		if (!cont) return typeid(void);
+		auto val = cont.get();
+		return typeid(*val);
+	}
 private:
```
Comment (Tyilo) [inline tpie/pipelining/container.h]: Yeah, fixed in 133dc02

File: tpie/pipelining/node.cpp (hunk 1 above)
Comment (Tyilo) [inline tpie/pipelining/node.cpp]: I think the node specific type is fine, but I have changed the other two to auto in 7d665c6

File: tpie/pipelining/node.h (hunk 2 above)
Comment (Tyilo) [inline tpie/pipelining/node.h]: I agree, it has been removed in 6782409

File: tpie/pipelining/node.cpp (hunk 3 above)
Comment (Tyilo) [inline tpie/pipelining/node.cpp]: I think it is clearer when using optional and it also allows you to use `value_or`

File: tpie/pipelining/tokens.h (hunk 4 above)
Comment (Tyilo) [inline tpie/pipelining/tokens.h]: I don't think I understand. Are you referring to `forward` or `fetch_maybe`?

Comment (antialize) [inline tpie/pipelining/tokens.h]: I do not understad who is supposed to call this forward, and fetch_maybe function, it must be called from somewhere?

Comment (Tyilo) [inline tpie/pipelining/tokens.h]: `fetch_maybe` is called from https://github.com/thomasmoelhave/tpie/blob/6782409ecb15fd386c79a2560d65b377329e9f12/tpie/pipelining/node.cpp#L235

`forward` is called from https://github.com/thomasmoelhave/tpie/blob/6782409ecb15fd386c79a2560d65b377329e9f12/tpie/pipelining/pipeline.cpp#L148

Comment (antialize) [inline tpie/pipelining/tokens.h]: I see, all is good then.

**Comment by Tyilo** [comment]
@antialize Ready to merge?

**Pull request merged**